### PR TITLE
Changes required for MCTS with Rules Paper

### PR DESCRIPTION
**Pull request by klemense1**
- Simulation of full Interaction Dataset in Closed loop
- Integration of rules-mcts

### CODE REVIEW

File: bark/models/dynamic/integration.hpp [hunk 1]
```diff
@@ -10,15 +10,21 @@
...
 
-inline State euler_int(const DynamicModel& model, const State& x,
-                       const Input& u, float dt) {
-  return x + dt * model.StateSpaceModel(x, u);
+inline State euler_int(const DynamicModel& model,
+                const State &x,
+                const Input &u,
+                float dt) {
+  State new_x = x + dt * model.StateSpaceModel(x, u);
+  new_x(StateDefinition::THETA_POSITION) =
```
Comment (patrickhart) [inline bark/models/dynamic/integration.hpp:25]: this should not be in the integration function. Why is this required now?

File: bark/models/execution/interpolation/interpolate.cpp [hunk 2]
```diff
@@ -82,9 +82,14 @@ State ExecutionModelInterpolate::Interpolate(const State& p0, const State& p1,
 void ExecutionModelInterpolate::Execute(const float& new_world_time,
                                         const Trajectory& trajectory,
                                         const DynamicModelPtr dynamic_model) {
+
   // book-keeping
   SetLastTrajectory(trajectory);
 
+  if(GetExecutionStatus() == ExecutionStatus::INVALID) {
+    return;
```
Comment (patrickhart) [inline bark/models/execution/interpolation/interpolate.cpp:90]: consider adding a log message

File: bark/models/tests/behavior_not_started_test.cc [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#include "bark/world/tests/make_test_world.hpp"
+
+using bark::models::behavior::BehaviorNotStarted;
+using bark::models::behavior::BehaviorStatus;
+using bark::world::Agent;
+using bark::world::ObservedWorld;
+using bark::world::World;
+using bark::world::WorldPtr;
+
+TEST(behavior_static_trajectory_plan, plan) {
```
Comment (patrickhart) [inline bark/models/tests/behavior_not_started_test.cc:23]: this test seems not much of a use

File: bark/python_wrapper/world/agent.cpp [hunk 4]
```diff
@@ -71,10 +73,11 @@ void python_agent(py::module m) {
                 a.GetExecutionModel(),                           // 7
                 a.GetDynamicModel(),                             // 8
                 a.GetCurrentState(),                             // 9
-                GoalDefinitionToPython(a.GetGoalDefinition()));  // 10
+                a.GetFirstValidTimestamp(),                      // 10
```
Comment (patrickhart) [inline bark/python_wrapper/world/agent.cpp:76]: Should this not be handled by the behavior model?

File: bark/runtime/commons/parameters.py [hunk 5]
```diff
@@ -56,6 +54,21 @@ def __getitem__(self, key):
                 self.store[new_key] = ParameterServer(log_if_default = self.log_if_default)
                 return self.store[new_key]
 
+    def __contains__(self, key):
+        return key in self.store
+
+    def AppendParamServer(self, p_server):
```
Comment (patrickhart) [inline bark/runtime/commons/parameters.py:60]: what does this do? Was this so far not possible via AddChild(..)?
@juloberno

Comment (patrickhart) [reply bark/runtime/commons/parameters.py:60]: Or we could consider handling this via the above named function

File: bark/runtime/scenario/interaction_dataset_processing/dataset_decomposer.py [hunk 6]
```diff
@@ -45,7 +46,7 @@ def __find_first_ts_on_map__(self, id_ego):
             for lane in lane_list:
                 polygon = self._map_interface.GetRoadgraph().GetLanePolygonForLaneId(lane.lane_id)
                 if Collide(polygon, point_agent):
-                    time_ego_first = state[0]*1e3  # use timestamp in ms
+                    time_ego_first = state[0]*1e3 + self._starting_offset_ms # use timestamp in ms
```
Comment (patrickhart) [inline bark/runtime/scenario/interaction_dataset_processing/dataset_decomposer.py:49]: is this not dataset specific? Could introduce a parameter for ms/s

File: bark/runtime/scenario/interaction_dataset_processing/interaction_dataset_reader.py [hunk 7]
```diff
@@ -66,16 +69,19 @@ def InitStateFromTrack(track, start):
...
-                              Point2d(1.5, 0)])
-    goal_polygon = goal_polygon.Translate(Point2d(bark_state[0, int(StateDefinition.X_POSITION)],
-                                                  bark_state[0, int(StateDefinition.Y_POSITION)]))
+    goal_polygon = Polygon2d(np.array([0.5 * goal_size, 0.5 * goal_size, 0.0]),
+                             [Point2d(0.0, 0.0),
+                              Point2d(goal_size, 0.0),
+                              Point2d(goal_size, goal_size),
+                              Point2d(0.0, goal_size),
+                              Point2d(0.0, 0.0)])
+    goal_polygon = goal_polygon.Translate(Point2d(bark_state[0, int(StateDefinition.X_POSITION)] - 0.5 * goal_size,
```
Comment (patrickhart) [inline bark/runtime/scenario/interaction_dataset_processing/interaction_dataset_reader.py:83]: format

File: bark/world/objects/agent.cpp [hunk 8]
```diff
@@ -16,6 +16,11 @@ namespace bark {
 namespace world {
 namespace objects {
 
+enum AgentStatus : unsigned int {
```
Comment (patrickhart) [inline bark/world/objects/agent.cpp:19]: as mentioned before, should we not handle that in the behavior model.
The behavior model sets INVALID, the agent does not move.
Seems redundant

File: bark/world/prediction/prediction_settings.cpp [hunk 9]
```diff
@@ -7,30 +7,70 @@
...
+                                  specific_prediction_agents.end()) {}
 
 void PredictionSettings::ApplySettings(
     bark::world::ObservedWorld& observed_world) const {
-  observed_world.SetEgoBehaviorModel(
-      BehaviorModelPtr(ego_prediction_model_->Clone()));
-  for (const auto& agent : observed_world.GetOtherAgents()) {
-    observed_world.SetBehaviorModel(
-        agent.first, BehaviorModelPtr(others_prediction_model_->Clone()));
+  // TODO: Clean up HACK to keep Behavior Status
```
Comment (patrickhart) [inline bark/world/prediction/prediction_settings.cpp:35]: now we need to specify several agent models?
what is specific_prediction_model_?
Could you clarify this?

File: bark/world/world.cpp [hunk 10]
```diff
@@ -61,17 +61,20 @@ void World::PlanAgents(const float& delta_time) {
   WorldPtr current_world(this->Clone());
   const float inc_world_time = world_time_ + delta_time;
   for (auto agent : agents_) {
-    ObservedWorld observed_world(current_world, agent.first);
-    agent.second->PlanBehavior(delta_time, observed_world);
-    if (agent.second->GetBehaviorStatus() == BehaviorStatus::VALID)
-      agent.second->PlanExecution(inc_world_time);
+    if (agent.second->IsValidAtTime(world_time_)) {
```
Comment (patrickhart) [inline bark/world/world.cpp:64]: who sets if it is valid?

File: bark/world/world.cpp [hunk 11]
```diff
@@ -61,17 +61,20 @@ void World::PlanAgents(const float& delta_time) {
...
+        agent.second->PlanExecution(inc_world_time);
+    }
   }
 }
 
 void World::Execute(const float& world_time) {
   using models::dynamic::StateDefinition::TIME_POSITION;
   for (auto agent : agents_) {
-    if (agent.second->GetBehaviorStatus() == BehaviorStatus::VALID &&
+    if (agent.second->IsValidAtTime(world_time_) &&
```
Comment (patrickhart) [inline bark/world/world.cpp:76]: not required as the behavior model would set invalid right?

**Review by patrickhart — CHANGES_REQUESTED** [review]
- long PR (50 files changed)
- decide on indent for formatting (python)
- I do not like the is valid at time for an agent. The behavior model knows whether there is a valid trajectory, also in case of the dataset right? Thus, if the behavior status is set to INVALID, the agent basically should not exist. It can be argued which of these two should be removed.
- can you elaborate the changes for the prediction setting?

**Comment by klemense1** [comment]
Prediction settings now allow to set behavior models for the ego vehicle, a default prediction model for the other vehicle, and a specific prediction for some agent ids (can be used e.g. to use a more detailed model for agents). The feature about the specific prediction model was added.

File: bark/benchmark/benchmark_runner.py [hunk 12]
```diff
@@ -24,6 +24,12 @@
 from bark.core.world.evaluation import *
 from bark.core.world.evaluation.ltl import *
 
+try:
+  from bark.core.world.evaluation.ltl import *
```
[the remark this answers is not in the document]
Comment (klemense1) [reply bark/benchmark/benchmark_runner.py:28]: right

File: bark/models/behavior/rule_based/lane_change_behavior.cpp [hunk 13]
```diff
@@ -138,8 +138,8 @@ BehaviorLaneChangeRuleBased::ChooseLaneCorridor(
       }
     }
     if (tmp_lane_corr != lane_corr) {
-      LOG(INFO) << "Agent " << observed_world.GetEgoAgentId()
-                << " is changing lanes." << std::endl;
+      VLOG(1) << "Agent " << observed_world.GetEgoAgentId()
```
[the remark this answers is not in the document]
Comment (klemense1) [reply bark/models/behavior/rule_based/lane_change_behavior.cpp:141]: it is more convenient to control the logging output (you just have more numbers :-) during debugging

[the remark this answers is not in the document]
Comment (klemense1) [reply bark/models/behavior/rule_based/lane_change_behavior.cpp:141]: but I can remove it if it is an issue

File: bark/models/dynamic/integration.hpp (hunk 1 above)
Comment (klemense1) [reply bark/models/dynamic/integration.hpp:25]: it is to make sure theta is between [0 ... 2pi]. why not here?

**Review by klemense1 — COMMENTED** [review]
I cleaned up the prediction settings

File: bark/runtime/scenario/interaction_dataset_processing/interaction_dataset_reader.py (hunk 7 above)
Comment (klemense1) [reply bark/runtime/scenario/interaction_dataset_processing/interaction_dataset_reader.py:83]: maybe I am doing something wrong but for my python formatter this file is okay. are we using the same formatting file (from bark?)

File: bark/world/objects/agent.cpp (hunk 8 above)
Comment (klemense1) [reply bark/world/objects/agent.cpp]: your're right, I removed it

File: bark/world/prediction/prediction_settings.cpp (hunk 9 above)
Comment (klemense1) [reply bark/world/prediction/prediction_settings.cpp]: Prediction settings now allow to set behavior models for the ego vehicle, a default prediction model for the other vehicle, and a specific prediction for some agent ids. The feature about the specific prediction model was added. It can be used for a more detailed model for nearby agents. Or to let the multi-agent mcts only plan for nearby agents, and predict the others with some default model

File: bark/models/behavior/not_started/behavior_not_started.hpp [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+namespace models {
+namespace behavior {
+
+using dynamic::Trajectory;
+using world::ObservedWorld;
+
+// model that always has behavior status not started (to be used in search,
+// replacing BehaviorStaticTrajectory, to prevent that BehaviorStaticTrajectory
+// becomes true in forward search and thus suddenly unexpectingly shows up)
+class BehaviorNotStarted : public BehaviorModel {
```
[the remark this answers is not in the document]
Comment (klemense1) [reply bark/models/behavior/not_started/behavior_not_started.hpp:29]: because it is not just data-replay, but closed loop simulation. and I am populating the scene with agents (some of them are ready at t=0, some of them later), but I do not want the agents to become valid during forward search. So for all that are not valid at t=t_planningstart, I replace them by BehaviorNotStarted for that planning loop

File: bark/models/behavior/not_started/BUILD [hunk 15]
```diff
@@ -0,0 +1,16 @@
+cc_library(
```
[the remark this answers is not in the document]
Comment (klemense1) [reply bark/models/behavior/not_started/BUILD:1]: behavior model that always has behavior status not started

File: bark/world/world.cpp (hunk 11 above)
Comment (klemense1) [reply bark/world/world.cpp:76]: I needed a separate mechanism to let agents start later that, as there might be a trajectory from the dataset, but it might not be within the map yet. if you want to start planning from there, that's a problem. 
Essentially, lifting the mechanism of checking validity to the agent level is much more powerful / convenient. We might consider removing the BehaviorStatus.

File: bark/world/world.cpp (hunk 10 above)
Comment (klemense1) [reply bark/world/world.cpp:64]: Agent has a function 
`bool Agent::IsValidAtTime(const float& world_time) const {
  return isgreaterequal(world_time + std::numeric_limits<float>::epsilon(),
                        first_valid_timestamp_);
}
`

File: bark/models/execution/interpolation/interpolate.cpp (hunk 2 above)
Comment (klemense1) [reply bark/models/execution/interpolation/interpolate.cpp:90]: done

File: bark/world/objects/agent.cpp (hunk 8 above)
Comment (klemense1) [reply bark/world/objects/agent.cpp]: I removed the enum

File: bark/python_wrapper/world/agent.cpp (hunk 4 above)
Comment (klemense1) [reply bark/python_wrapper/world/agent.cpp:76]: wasn't sufficient. We might consider removing the BehaviorStatus

File: bark/runtime/scenario/interaction_dataset_processing/dataset_decomposer.py (hunk 6 above)
Comment (klemense1) [reply bark/runtime/scenario/interaction_dataset_processing/dataset_decomposer.py:49]: sure, I introduced a parameter

File: bark/world/world.cpp (hunk 10 above)
Comment (klemense1) [reply bark/world/world.cpp:64]: default is first_valid_timestamp_ = 0.0
when simulating the dataset in closed loop, these agents get different first_valid_timestamps.
everything else should work as usual

File: bark/models/dynamic/integration.hpp (hunk 1 above)
Comment (patrickhart) [reply bark/models/dynamic/integration.hpp:25]: It somehow should not be in the integration function. A normalization function could be added. 
But also okay for me

File: bark/runtime/scenario/interaction_dataset_processing/dataset_decomposer.py (hunk 6 above)
Comment (patrickhart) [reply bark/runtime/scenario/interaction_dataset_processing/dataset_decomposer.py]: But also the 1e3?

File: bark/runtime/scenario/interaction_dataset_processing/interaction_dataset_reader.py (hunk 7 above)
Comment (patrickhart) [reply bark/runtime/scenario/interaction_dataset_processing/interaction_dataset_reader.py:83]: 80 chars per line, is that not set currently?

File: bark/world/prediction/prediction_settings.cpp (hunk 9 above)
Comment (patrickhart) [reply bark/world/prediction/prediction_settings.cpp]: okay, but should we not have a constructor for the case we do not need a specific prediction  model?
E.g. overload the constructor

File: bark/world/world.cpp (hunk 11 above)
Comment (patrickhart) [reply bark/world/world.cpp:76]: yeah, maybe we should discuss this. From my perspective I would not have a problem with dropping the behavior status then

File: bark/runtime/scenario/interaction_dataset_processing/interaction_dataset_reader.py (hunk 7 above)
Comment (klemense1) [reply bark/runtime/scenario/interaction_dataset_processing/interaction_dataset_reader.py]: okay, I formatted it

File: bark/world/prediction/prediction_settings.cpp (hunk 9 above)
Comment (klemense1) [reply bark/world/prediction/prediction_settings.cpp]: I implemented it

**Review by patrickhart — APPROVED** [review]
